src/main/python/gemini/beam_to_pydss_federate.py

```diff
@@ -0,0 +1,147 @@
+# this file creates an intermediate federate
+# it maps the coordinates from TEMPO to the nearest charging
+# station modeled in PyDSS
+import time
+import helics as h
+import numpy as np
+import logging
+import json
+import os
+
+
+def run_beam_to_pydss_federate(station_bus_pairs):
+    logging.basicConfig(filename='beam_to_pydss_federate.log', level=logging.DEBUG, filemode='w')
+    fedinfo = h.helicsCreateFederateInfo()
+
+    # set the name
+    h.helicsFederateInfoSetCoreName(fedinfo, "beam_to_pydss_federate")
+
+    # set core type
+    h.helicsFederateInfoSetCoreTypeFromString(fedinfo, "zmq")
+
+    # set initialization string
+    h.helicsFederateInfoSetCoreInitString(fedinfo, "--federates=1")
+
+    # set message interval
+    deltat = 1.0  # smallest discernable interval to this federate
+    h.helicsFederateInfoSetTimeProperty(fedinfo, h.helics_property_time_delta, deltat)
+
+    # create federate
+    cfed = h.helicsCreateCombinationFederate("beam_to_pydss_federate", fedinfo)
+    print("beam_to_pydss_federate created")
+    logging.info("beam_to_pydss_federate created")
+
+    # register publications
+    # publish an ordered list of charging station codes in same order as charging loads
+    # pubs_station_loads = {}
+    #
+    # for s in range(len(station_bus_pairs)):
+    #     station_id = station_bus_pairs[s][0]
+    #     pubs_station_loads[station_id] = h.helicsFederateRegisterTypePublication(cfed, station_id, "string_vector", "")
+
+
+    # Register a publication of control signals
+    pubs_control = h.helicsFederateRegisterTypePublication(cfed, "pubs_power_limit_and_lpm_control", "string", "")
+
+    print("publications registered")
+
+    # register subscriptions
+    # subscribe to information from TEMPO such that you can map to PyDSS modeled charging stations
+    subs_charger_loads = h.helicsFederateRegisterSubscription(cfed, "beamFederate/chargingLoad", "string")
+    print("subscriptions registered")
+
+    # enter execution mode
+    h.helicsFederateEnterExecutingMode(cfed)
+    print("beam_to_pydss_federate in execution mode")
+    logging.info("beam_to_pydss_federate in execution mode")
+
+    currenttime = -1
+    timebin = 300
+    # start execution loop
+    for t in range(0, timebin*60*12+1, timebin):
+        while currenttime < t:
+            currenttime = h.helicsFederateRequestTime(cfed, t)
+        isupdated = 0
+        while isupdated != 1:
+            isupdated = h.helicsInputIsUpdated(subs_charger_loads)
+        print(f"charger loads received at currenttime: {t} seconds")
```

Comment:
would it be better to insert time.sleep in each loop to decrease CPU consumption?

Reply:
We will figure it out once we have the PyDSS running. let's keep it this way for now.